### PR TITLE
kubectl exec/attach using websocket

### PR DESCRIPTION
**Pull request by robberphex**
#### Which issue(s) this PR fixes:

Implements #89163

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by robberphex** [comment]
I signed it

**Comment by robberphex** [comment]
I signed it

**Comment by aojea** [comment]
current commit history is not easy to review, it seems it maintains the order used for local development but not for code review.

Also, the BUILD files are no longer needed

**Comment by robberphex** [comment]
> current commit history is not easy to review, it seems it maintains the order used for local development but not for code review.
> 
> Also, the BUILD files are no longer needed

I was hoping to keep commits from `mlbiam <marc@tremolo.io>`, so the commit log is not good to review. Fixed this, and removed useless BUILD file.

File: staging/src/k8s.io/apimachinery/pkg/util/httpstream/websocket/connection.go [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+	streamOut *io.PipeWriter
+}
+
+// CreateStream does nothing as httpstream.Stream is a SPDY function, not a websocket concept
+func (connection *Connection) CreateStream(headers http.Header) (httpstream.Stream, error) {
+	return nil, nil
+}
+
+// CreateStream creates a stream by the channel id
+func (connection *Connection) CreateStrean(channelID int) (httpstream.Stream, error) {
```
Comment (aojea) [inline staging/src/k8s.io/apimachinery/pkg/util/httpstream/websocket/connection.go:51]: is this a typo? I can't find it in other place

**Comment by aojea** [comment]
> I was hoping to keep commits from `mlbiam <marc@tremolo.io>`, so the commit log is not good to review. Fixed this, and removed useless BUILD file.

sorry, I didn't mean to squash all, but to have commits that were reviewable, previously they were containing comments and code that was being discarded between commits

File: staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go [hunk 2]
```diff
@@ -0,0 +1,479 @@
...
+		// before returning. When hijack finishes copying stdout/stderr, it calls
+		// Close() on its side of remoteStdin, which allows this copy to complete.
+		// When that happens, we must Close() on our side of remoteStdin, to
+		// allow the copy in hijack to complete, and hijack to return.
+		go func() {
+			defer runtime.HandleCrash()
+			defer once.Do(func() { p.remoteStdinIn.Close() })
+
+			// this "copy" doesn't actually read anything - it's just here to wait for
+			// the server to close remoteStdin.
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go:114]: Isn't `remosteStdinIn` part of a local pipe? Why would server close it?

File: staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go [hunk 3]
```diff
@@ -0,0 +1,479 @@
...
+	}()
+}
+
+func (p *streamProtocolV4) stream(conn *websocket.Conn) error {
+
+	defer conn.Close()
+	doneChan := make(chan struct{}, 2)
+
+	// set up error stream
+	p.errorStreamIn, p.errorStreamOut = io.Pipe()
```
Comment (caesarxuchao) [inline staging/src/k8s.io/client-go/tools/remotecommandwebsocket/v4.go:165]: Why do we need these pipes? Can we stitch together the websocket connection with the streamOption's stdin/stdout/stderr directly?

**Review by caesarxuchao — COMMENTED** [review]
Thanks for reviving this, @robberphex.

Some high-level thoughts:
* We implemented a dumb httpstream.Connection interface which required a lot of boilerplate. But `httpstream.Connection` is SPDY specific, which supports multiplexing "streams" on the same "connection". WebSocket doesn't have a native multiplexing support. We should consider giving up the `httpstream.Connection` interface when implementing WebSocket.

* the remotecommandwebsocket/v4.go needs some clean up. I'm not sure why we need to create a set of intermediary pipes there. Also we should consider if we can deduplicate remotecommandwebsocket and the existing remotecommand package.

* We need to incorporate the latest executor/spdy improvements (#105632, #103177, and perhaps #107605) into the websocket implementation.

I haven't looked at the unit tests yet.

**Comment by caesarxuchao** [comment]
cc @dberkov

**Comment by caesarxuchao** [comment]
/assign @jpbetz @yliaog 

in case I don't have bandwidth to review this PR in the future.

**Comment by enj** [comment]
There is ongoing [discussion](https://groups.google.com/g/kubernetes-sig-api-machinery/c/KKAEZMIJn5U/m/gB1fQLmbAgAJ) around this topic and it is clear that such a change will require a KEP to move forward.  @robberphex please collaborate with @ash2k on next steps.

/close

**Pull request closed without merging**